sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala

```diff
@@ -739,15 +739,24 @@ object HashJoin extends CastSupport with SQLConfHelper {
     // jump over keys that have a higher index value than the required key
     if (keys.size == 1) {
       assert(index == 0)
-      cast(BoundReference(0, LongType, nullable = false), keys(index).dataType)
+      Cast(
+        child = BoundReference(0, LongType, nullable = false),
+        dataType = keys(index).dataType,
+        timeZoneId = Option(conf.sessionLocalTimeZone),
```

Comment:
It seems that `rewriteKeyExpr()` above does not support timezone-related data type yet (Line 716). Shall we add an assertion for keys data type to be `IntegralType` and sum of key sizes no larger that 8 bytes in this method?

Reply:
Sure, added assertion.

Reply:
I am keepting the `timeZoneId` here since we may support DateType/TimestampType in the future